app/updatemanager.go

```diff
@@ -125,9 +138,137 @@ func (u *UpdateManager) run(ctx context.Context) error {
 	return nil
 }
 
+type UpdateControlMapState struct {
+	Action           string `json:"action"`
+	OnMapExpire      string `json:"on_map_expire"`
+	OnActionExecuted string `json:"on_action_executed"`
+}
+
 type UpdateControlMap struct {
-	ID       string `json:"id"`
-	Priority int    `json:"priority"`
+	ID       string                           `json:"id"`
+	Priority int                              `json:"priority"`
+	States   map[string]UpdateControlMapState `json:"states"`
+}
+
+var UpdateControlMapStateKeyValid []string = []string{
+	"ArtifactInstall_Enter",
+	"ArtifactReboot_Enter",
+	"ArtifactCommit_Enter",
+}
+var UpdateControlMapStateActionValid []string = []string{
+	"continue",
+	"force_continue",
+	"pause",
+	"fail",
+}
+var UpdateControlMapStateOnMapExpireValid []string = []string{
+	"continue",
+	"force_continue",
+	"fail",
+}
+var UpdateControlMapStateOnActionExecutedValid []string = UpdateControlMapStateActionValid
+
+var UpdateControlMapStateActionDefault string = "continue"
+
+func UpdateControlMapStateOnMapExpireDefault(action string) string {
+	if action == "pause" {
+		return "fail"
+	}
+	return action
+}
+
+var UpdateControlMapStateOnActionExecutedDefault = UpdateControlMapStateActionDefault
+
+func (s UpdateControlMapState) Validate() error {
```

Comment:
Just thinking:
Maybe we should also: https://golang.org/pkg/encoding/json/#Decoder.DisallowUnknownFields
In case of spelling errors, etc? :thinking:

Reply:
Thanks. I didn't know about this. I am reading up and adding it to the PR, it is a good idea